plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2536,9 +2638,235 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
+   EOS_ASSERT(_ro_trx_exec_tasks_fut.empty(),  producer_exception, "_ro_trx_exec_tasks_fut expected to be empty");
+
+   _ro_write_window_timer.cancel();
+   _ro_read_window_timer.cancel();
+
+   // we are in write window, so no read-only trx threads are processing transactions.
+   // _ro_trx_queue is not being accessed. No need to lock.
+   if ( _ro_trx_queue.queue.empty() ) { // no read-only trxs to process. stay in write window
+      start_write_window(); // restart write window timer for next round
+      return;
+   }
+
+   app().executor().set_to_read_window();
+   chain_plug->chain().set_db_read_only_mode();
+   _received_block = false;
+
+   // start read-only transaction execution tasks in the thread pool
+   _ro_num_active_trx_exec_tasks = _ro_thread_pool_size;
+   for (auto i = 0; i < _ro_thread_pool_size; ++i ) {
+      _ro_trx_exec_tasks_fut.emplace_back( post_async_task( _ro_thread_pool.get_executor(), [self = this] () {
+         return self->read_only_trx_execution_task();
+      }) );
+   }
+
+   auto expire_time = boost::posix_time::microseconds(_ro_read_window_time_us.count());
+   _ro_read_window_timer.expires_from_now( expire_time );
+   _ro_read_window_timer.async_wait( app().executor().wrap(
+      priority::high,
+      exec_queue::read_only_trx_safe,
+      [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
+         auto self = weak_this.lock();
+         if( self && ec != boost::asio::error::operation_aborted ) {
+            // use future make sure all read-only tasks finished before swithching to write window
+            for ( auto& task: self->_ro_trx_exec_tasks_fut )
+               task.get();
+            self->_ro_trx_exec_tasks_fut.clear();
+            self->switch_to_write_window();
+          } else if ( self )
+             self->_ro_trx_exec_tasks_fut.clear();
+       }));
+}
+
+bool producer_plugin_impl::read_only_trx_execution_task() {
+   auto start = fc::time_point::now();
+   auto read_window_deadline = start + _ro_read_window_effective_time_us;
+   while ( fc::time_point::now() < read_window_deadline && !_received_block ) {
+      std::unique_lock<std::mutex> lck( _ro_trx_queue.mtx );
+      if ( _ro_trx_queue.queue.empty() ) {
+         break;
+      }
+      auto trx = _ro_trx_queue.queue.front();
+      _ro_trx_queue.queue.pop_front();
+      lck.unlock();
+      
+      auto block_deadline_exhausted = process_read_only_transaction( trx.trx, trx.next, start );
+
+      // If a transaction passes deadline, that indicates we are close to the end of
+      // read window. Do not schedule new transactions
+      if ( block_deadline_exhausted ) {
+         lck.lock();
+         _ro_trx_queue.queue.push_front(trx);
+         break;
+      }
+   }
+
+   // If all tasks are finished, do not wait until end of read window; switch to write window now.
+   if ( --_ro_num_active_trx_exec_tasks == 0 ) {
+      _ro_trx_exec_tasks_fut.clear();
```

Comment:
It is not safe to do this here. `_ro_trx_exec_tasks_fut` is accessed by the timer lambda in `switch_to_read_window` which can run at the same time. Looks like this is not needed here anyway.

Reply:
Thanks. I think we still need it because we have an assert of its empty when starting switching to read window and it needs to be empty before emplacing new tasks. But we should do the clear inside the post so it's done on app thread.